src/modules/gatsby-plugin/gatsby-node.ts

```diff
@@ -109,174 +128,212 @@ const setupImgixClient = ({
+  let validatedOptions;
+  try {
+    validatedOptions = decodeOptions(_options);
+  } catch (error) {
+    const errorString = `[@imgix/gatsby] Fatal error during setup, plugin options are not valid: ${String(
+      error,
+    )}`;
+    gatsbyContext.reporter.panic(errorString);
+    throw new Error(errorString);
+  }
 
-      .let('packageVersion', VERSION)
-      .letL('imgixClient', ({ options, packageVersion }) =>
-        setupImgixClient({ options, packageVersion }),
-      )
-      .letL(
-        'typesAndFields',
-        ({ imgixClient, options: { defaultImgixParams } }) =>
-          buildImgixGatsbyTypes<{ rawURL: string }>({
-            cache: gatsbyContext.cache,
-            imgixClient,
-            resolveUrl: prop('rawURL'),
-            defaultParams: defaultImgixParams,
-          }),
-      )
-      .letL('imgixImageType', ({ typesAndFields }) =>
-        gatsbyContext.schema.buildObjectType({
-          name: 'ImgixImage',
-          fields: {
-            url: typesAndFields.fields.url,
-            fluid: typesAndFields.fields.fluid,
-            fixed: typesAndFields.fields.fixed,
-            gatsbyImageData: typesAndFields.fields.gatsbyImageData,
+  // This creates the "imgix client" (it is actually a modified version of the
+  // core imgix client), which will be passed around to be used in the
+  // application
+  const imgixClient = setupImgixClient({
+    options: validatedOptions,
+    packageVersion: VERSION,
+  });
+
+  // Create the imgix GraphQL types, which will be added to the schema later
+  const typesAndFields = buildImgixGatsbyTypes<{ rawURL: string }>({
+    cache: gatsbyContext.cache,
+    imgixClient,
+    resolveUrl: prop('rawURL'),
+    defaultParams: validatedOptions.defaultImgixParams,
+  });
+
+  // Create the root "imgixImage" type which will exist at the root of the
+  // GraphQL Query type, and also on every type that the user has specified to
+  // modify
+  const imgixImageType = gatsbyContext.schema.buildObjectType({
+    name: 'ImgixImage',
+    fields: {
+      url: typesAndFields.fields.url,
+      fluid: typesAndFields.fields.fluid,
+      fixed: typesAndFields.fields.fixed,
+      gatsbyImageData: typesAndFields.fields.gatsbyImageData,
+    },
+  });
+
+  const optionsFields = validatedOptions.fields ?? [];
+
+  const fieldTypes = createFieldTypes(
+    optionsFields,
+    gatsbyContext,
+    imgixImageType,
+  );
+  const rootType = createRootObjectType(gatsbyContext, imgixImageType);
+
+  // Now, we actually add the types and fields to the schema
+  try {
+    const { createTypes } = gatsbyContext.actions;
+    createTypes(typesAndFields.types.map(gatsbyContext.schema.buildObjectType));
+    createTypes(
+      typesAndFields.enumTypes.map(gatsbyContext.schema.buildEnumType),
+    );
+    createTypes(
+      typesAndFields.inputTypes.map(gatsbyContext.schema.buildInputObjectType),
+    );
+    createTypes(fieldTypes);
+    createTypes(imgixImageType);
+    createTypes(rootType);
```

Comment:
It would be great to get some more detail on what's going on here or even some links to documentation for reference.

Reply:
Sure thing. I've added some links to the docs, it's hard to explain what's going on here without going super in depth.